Source/CSharpEssentials/Extensions.cs

```diff
@@ -168,9 +170,64 @@ private static bool IsGeneratedCode(string filePath)
             return false;
         }
 
-        public static bool IsGeneratedCode(this SyntaxTree tree)
+        public static bool IsGeneratedCode(this SyntaxNode node)
         {
-            return IsGeneratedCode(tree.FilePath);
+            return IsInGeneratedCodeContext(node)
+                || IsGeneratedCode(node.SyntaxTree.FilePath);
+        }
+
+        private static bool IsInGeneratedCodeContext(this SyntaxNode node)
+        {
+            do
+            {
+                bool inGeneratedCodeContext = false;
+                switch (node.Kind())
+                {
+                    case SyntaxKind.ClassDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((ClassDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.StructDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((StructDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.MethodDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((MethodDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.PropertyDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((PropertyDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.IndexerDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((IndexerDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.EventDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((EventDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.FieldDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((FieldDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.OperatorDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((OperatorDeclarationSyntax)node).AttributeLists);
+                        break;
+                    case SyntaxKind.ConversionOperatorDeclaration:
+                        inGeneratedCodeContext = ContainsGeneratedCodeAttribute(((ConversionOperatorDeclarationSyntax)node).AttributeLists);
+                        break;
+
+                }
+
+                if (inGeneratedCodeContext)
+                    return true;
+
+                node = node.Parent;
+            }
+            while (node != null);
+
+            return false;
+        }
+
+        private static bool ContainsGeneratedCodeAttribute(IEnumerable<AttributeListSyntax> attributes)
```

Comment:
For performance, it's better to pass these attribute lists as ```SyntaxList<AttributeListSyntax>``` to avoid boxing the parameter (via the conversion to IEnumerable<T>)
Although, you'll immediately run into boxing again for the subsequent LINQ expression unless you expand it out by hand, e.g.
```C#
    foreach (var attributeList in attributes)
    {
        foreach (var x in attributeList.Attributes)
        {
            if (IsGeneratedCodeAttribute(x)) return true;
        }
    }

    return false;
```